### DISCUSSION TITLE
Python 3.10 keg only vs $PATH

### DISCUSSION THREAD

**Discussion by msarrel** [Homebrew/discussions#2730] [Everyday usage] [answered]
I installed python@3.10.  Since it is still keg-only, the instructions suggest this work around.  However, this fails for me.  Can anyone suggest an alternative until python@3.10 is no longer keg-only?
```
If you need to have python@3.10 first in your PATH, run:
  echo 'export PATH="/usr/local/opt/python@3.10/bin:$PATH"' >> /Users/msarrel/.bash_profile
```
I think the problem is the `@` in the path.  That seems to confuse `bash`.   Even just from the terminal, if I try to do file-name completion, `bash` gets confused.

Case 1:  So far so good.
```
MT-307040:bin msarrel$ ls /usr/local/opt/python
python/      python3/     python@3/    python@3.10/ python@3.9/
```

Case 2: Gets confused by unquoted `@`.
```
MT-307040:bin msarrel$ ls /usr/local/opt/python@
@::1                          @bigfix-sds-r06.jpl.nasa.gov  @bigfix-sds-r12.jpl.nasa.gov
@bigfix-sds-r01.jpl.nasa.gov  @bigfix-sds-r07.jpl.nasa.gov  @bigfix-sds-r13.jpl.nasa.gov
@bigfix-sds-r02.jpl.nasa.gov  @bigfix-sds-r08.jpl.nasa.gov  @bigfix-sds-r14.jpl.nasa.gov
@bigfix-sds-r03.jpl.nasa.gov  @bigfix-sds-r09.jpl.nasa.gov  @bigfix-sds.jpl.nasa.gov
@bigfix-sds-r04.jpl.nasa.gov  @bigfix-sds-r10.jpl.nasa.gov  @broadcasthost
@bigfix-sds-r05.jpl.nasa.gov  @bigfix-sds-r11.jpl.nasa.gov  @localhost
```

Case 3: Putting a `\` before the `@` seems to fix the problem.
```
MT-307040:bin msarrel$ ls /usr/local/opt/python\@3
python@3/    python@3.10/ python@3.9/  
```
I tried putting various numbers of `\`s in front of the `@` in my `.bash_profile`, but couldn't make it work.  I always get stuff like this:
```
MT-307040:~ msarrel$ echo $PATH
/usr/local/opt/python\\@3.10/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin
MT-307040:~ msarrel$ which python3
/usr/local/bin/python3
```
If it makes a difference, I'm using MacOs 11.6.2 Big Sur, and GNU bash, version 5.1.16(1)-release (x86_64-apple-darwin20.6.0).

**Comment by gromgit** [comment, marked as answer]
> Even just from the terminal, if I try to do file-name completion, `bash` gets confused.

That is a bash completions issue, not a `$PATH` issue. Try `brew install bash-completion` if you're using the standard macOS bash, or `brew install bash-completion@2` if you're using the Homebrew bash. Don't forget to follow the instructions at the end of the install in either case.
